packages/atlas-service/src/main.ts

```diff
@@ -124,6 +142,87 @@ export class AtlasService {
+    this.attachOidcPluginLoggerEvents();
+  }
+
+  private static attachOidcPluginLoggerEvents() {
+    // NB: oidc-plugin will emit these log events for all auth states that it
+    // keeps and it doesn't provide any information for us to distinguish
+    // between them. This is okay to ignore for our case for now as we know that
+    // only one auth state is being part of oidc-plugin state
+    this.oidcPluginLogger.on('mongodb-oidc-plugin:refresh-started', () => {
+      this.oidcPluginSyncedFromLoggerState = 'expired';
+    });
+    this.oidcPluginLogger.on('mongodb-oidc-plugin:refresh-failed', () => {
+      this.token = null;
+      this.oidcPluginSyncedFromLoggerState = 'error';
+      this.oidcPluginLogger.emit('atlas-service-token-refresh-failed');
+    });
+    // NB: oidc-plugin only has a logger interface to listen to state updates,
+    // it also doesn't expose renewed tokens or any other state in any usable
+    // way from those events or otherwise, so to get the renewed token we listen
+    // to the refresh-succeeded event and then kick of the "refresh"
+    // programmatically to be able to get the actual tokens back and sync them
+    // to the service state
+    let refreshing = false;
+    this.oidcPluginLogger.on('mongodb-oidc-plugin:refresh-succeeded', () => {
+      // In case our call to REFRESH_TOKEN_CALLBACK somehow started another
+      // refresh token instead of just returning token form the plugin state, we
+      // short circuit if plugin logged a refresh-succeeded event and this
+      // listener got triggered
+      if (refreshing) {
+        return;
+      }
+      refreshing = true;
+      void this.plugin.mongoClientOptions.authMechanismProperties
+        // WARN: in the oidc plugin refresh callback is actually the same method
+        // as sign in, so calling it here means that potentially this can start
+        // an actual sign in flow for the user instead of just trying to refresh
+        // the token
+        .REFRESH_TOKEN_CALLBACK(
+          { clientId: this.clientId, issuer: this.issuer },
+          {
+            // Required driver specific stuff
+            version: 0,
+          }
+        )
+        .then((token) => {
+          this.token = token;
+          this.oidcPluginSyncedFromLoggerState = 'authenticated';
+          this.oidcPluginLogger.emit('atlas-service-token-refreshed');
+        })
+        .catch(() => {
+          this.token = null;
+          this.oidcPluginSyncedFromLoggerState = 'error';
+          this.oidcPluginLogger.emit('atlas-service-token-refresh-failed');
+        })
+        .finally(() => {
+          refreshing = false;
+        });
+    });
+  }
+
+  private static async maybeWaitForToken({
+    signal,
+  }: { signal?: AbortSignal } = {}) {
+    if (signal?.aborted) {
+      return;
+    }
+    // In cases where we ended up in expired state, we know that oidc-plugin
+    // will try to refresh the token automatically, we can wait for this process
```

Comment:
It's already doing it if we set our state to `expired`

```suggestion
    // is trying to refresh the token automatically, we can wait for this process
```